File: backend/routers/rooms.py
"""Room management routes for CRUD operations and pause/resume."""

from typing import List

import crud
import schemas
from auth import require_admin
from database import get_db
from dependencies import (
    RequestIdentity,
    ensure_room_access,
    get_agent_manager,
    get_chat_orchestrator,
    get_request_identity,
)
from exceptions import RoomAlreadyExistsError
from fastapi import APIRouter, Depends, HTTPException
from orchestration import ChatOrchestrator
from sdk import AgentManager
from services.agent_service import clear_room_messages_with_cleanup, delete_room_with_cleanup
from sqlalchemy.ext.asyncio import AsyncSession

router = APIRouter()


@router.get("", response_model=List[schemas.RoomSummary])
async def list_rooms(identity: RequestIdentity = Depends(get_request_identity), db: AsyncSession = Depends(get_db)):
    """List all chat rooms."""
    rooms = await crud.get_rooms(db, identity)
    return rooms


@router.post("", response_model=schemas.Room)
async def create_room(
    room: schemas.RoomCreate,
    identity: RequestIdentity = Depends(get_request_identity),
    db: AsyncSession = Depends(get_db),
):
    """Create a new chat room."""
    try:
        owner_id = "admin" if identity.role == "admin" else identity.user_id

        # Explicitly check for duplicate chat-mode room names (rooms without world_id)
        # Since world_id is NULL for chat rooms, SQL NULL != NULL doesn't catch duplicates
        existing_rooms = await crud.get_rooms(db, identity)
        for existing_room in existing_rooms:
            if existing_room.name == room.name:
                raise RoomAlreadyExistsError(room.name)

        return await crud.create_room(db, room, owner_id=owner_id)
    except RoomAlreadyExistsError:
        raise
    except Exception as e:
        error_message = str(e)
        if (
            "UNIQUE constraint failed: rooms.owner_id, rooms.name" in error_message
            or "UNIQUE constraint failed: rooms.name" in error_message
            or "UNIQUE constraint failed: rooms.owner_id, rooms.name, rooms.world_id" in error_message
        ):
            raise RoomAlreadyExistsError(room.name)
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{room_id}", response_model=schemas.Room)
async def get_room(
    room_id: int,
    identity: RequestIdentity = Depends(get_request_identity),
    db: AsyncSession = Depends(get_db),
):
    """Get a specific room by ID."""
    return await ensure_room_access(db, room_id, identity)


@router.patch("/{room_id}", response_model=schemas.Room)
async def update_room(
    room_id: int,
    room_update: schemas.RoomUpdate,
    identity: RequestIdentity = Depends(get_request_identity),
    db: AsyncSession = Depends(get_db),
):
    """Update room configuration (max_interactions, is_paused)."""
    await ensure_room_access(db, room_id, identity)
    room = await crud.update_room(db, room_id, room_update)
    if room is None:
        raise HTTPException(status_code=404, detail="Room not found")
    return room


@router.post("/{room_id}/pause", response_model=schemas.Room)
async def pause_room(
    room_id: int,
    db: AsyncSession = Depends(get_db),
    identity: RequestIdentity = Depends(get_request_identity),
    agent_manager: AgentManager = Depends(get_agent_manager),
    chat_orchestrator: ChatOrchestrator = Depends(get_chat_orchestrator),
):
    """
    Pause agent interactions in a room and interrupt any active agent responses.
    """
    await ensure_room_access(db, room_id, identity)

    room_update = schemas.RoomUpdate(is_paused=True)
    room = await crud.update_room(db, room_id, room_update)
    if room is None:
        raise HTTPException(status_code=404, detail="Room not found")

    # Interrupt any agents currently processing responses in this room
    # Pass db to save any partial responses that were in-progress
    await chat_orchestrator.interrupt_room_processing(room_id, agent_manager, db=db)

    return room


@router.post("/{room_id}/resume", response_model=schemas.Room)
async def resume_room(
    room_id: int,
    identity: RequestIdentity = Depends(get_request_identity),
    db: AsyncSession = Depends(get_db),
):
    """Resume agent interactions in a room."""
    await ensure_room_access(db, room_id, identity)
    room_update = schemas.RoomUpdate(is_paused=False)
    room = await crud.update_room(db, room_id, room_update)
    if room is None:
        raise HTTPException(status_code=404, detail="Room not found")
    return room


@router.delete("/{room_id}", dependencies=[Depends(require_admin)])
async def delete_room(
    room_id: int,
    db: AsyncSession = Depends(get_db),
    agent_manager: AgentManager = Depends(get_agent_manager),
    chat_orchestrator: ChatOrchestrator = Depends(get_chat_orchestrator),
):
    """Delete a room and cleanup all associated agent clients. (Admin only)"""
    success = await delete_room_with_cleanup(db, room_id, agent_manager, chat_orchestrator)
    if not success:
        raise HTTPException(status_code=404, detail="Room not found")
    return {"message": "Room deleted successfully"}


@router.delete("/{room_id}/messages", dependencies=[Depends(require_admin)])
async def clear_room_messages(
    room_id: int,
    db: AsyncSession = Depends(get_db),
    agent_manager: AgentManager = Depends(get_agent_manager),
    chat_orchestrator: ChatOrchestrator = Depends(get_chat_orchestrator),
):
    """Clear all messages from a room and reset agent sessions. (Admin only)"""
    success = await clear_room_messages_with_cleanup(db, room_id, agent_manager, chat_orchestrator)
    if not success:
        raise HTTPException(status_code=404, detail="Room not found or no messages to delete")

    return {"message": "All messages cleared successfully"}

File: backend-ts/src/http/routes/rooms/rooms.ts
/**
 * Chat room CRUD — port of `backend/routers/rooms.py`.
 *
 * Mounted at the root; every path here is written out in full. The collection
 * routes are registered on both `/rooms` and `/rooms/` for the reason spelled
 * out in `routes/game/index.ts`: Python's canonical path carries the trailing
 * slash and Starlette redirects onto it, while Hono is strict and does not.
 *
 * The 404 strings differ from handler to handler and are Python's verbatim —
 * `ensure_room_access` raises `RoomNotFoundError` ("Room with id N not found")
 * while the handlers' own guards say "Room not found". The frontend renders
 * `detail`, so the inconsistency is observable and is reproduced, not tidied.
 */

import { Hono, type Context } from 'hono'

import {
  createRoom as createRoomRow,
  getRooms,
  updateRoom as updateRoomRow,
} from '../../../crud/rooms'
import { RoomAlreadyExistsError } from '../../../domain/errors'
import { getLogger } from '../../../infrastructure/logging/logger'
import { RoomCreate, RoomUpdate, toRoom, toRoomSummary } from '../../../schemas/rooms'
import { HttpError } from '../../errors'
import { requireAdmin } from '../../middleware/auth'
import { identityOf, type AppState } from '../../state'
import type { AppEnv } from '../../types'
import { intPathParam, parseBody } from '../game/shared'
import { ensureRoomAccessFor } from './shared'

const logger = getLogger('RoomRouter')

/** Python's guard string, shared by update, pause and resume. */
function roomNotFound(): HttpError {
  return new HttpError(404, 'Room not found')
}

export function createRoomRoutes(state: AppState): Hono<AppEnv> {
  const routes = new Hono<AppEnv>()

  // ---------------------------------------------------------------------------
  // Collection
  // ---------------------------------------------------------------------------

  function listRooms(c: Context<AppEnv>): Response {
    return c.json(getRooms(state.db, identityOf(c)).map(toRoomSummary))
  }

  async function createRoom(c: Context<AppEnv>): Promise<Response> {
    const body = await parseBody(c, RoomCreate)
    const identity = identityOf(c)
    const ownerId = identity.role === 'admin' ? 'admin' : identity.userId

    // Checked before the insert rather than caught after it, and this is
    // deliberate: a chat room has `world_id = NULL`, and the unique index that
    // would catch a duplicate is over (owner_id, name, world_id) — where SQL's
    // NULL != NULL means two chat rooms with the same name never collide.
    // Python does the same pre-check for exactly this reason (`rooms.py:44`).
    for (const existing of getRooms(state.db, identity)) {
      if (existing.name === body.name) throw new RoomAlreadyExistsError(body.name)
    }

    try {
      const created = createRoomRow(
        state.db,
        { name: body.name, maxInteractions: body.max_interactions },
        ownerId,
      )
      return c.json(toRoom(created))
    } catch (error) {
      // The pre-check races: two requests can both pass it. The index still
      // holds, so a constraint error becomes the same 409 rather than a 500.
      const message = error instanceof Error ? error.message : String(error)
      if (message.includes('UNIQUE constraint failed') && message.includes('rooms.')) {
        throw new RoomAlreadyExistsError(body.name)
      }
      throw error
    }
  }

  routes.get('/rooms', listRooms)
  routes.get('/rooms/', listRooms)
  routes.post('/rooms', createRoom)
  routes.post('/rooms/', createRoom)

  // ---------------------------------------------------------------------------
  // Single room
  // ---------------------------------------------------------------------------

  routes.get('/rooms/:room_id', (c) =>
    c.json(toRoom(ensureRoomAccessFor(c, state.db, intPathParam(c, 'room_id')))),
  )

  routes.patch('/rooms/:room_id', async (c) => {
    const roomId = intPathParam(c, 'room_id')
    ensureRoomAccessFor(c, state.db, roomId)

    const body = await parseBody(c, RoomUpdate)
    const updated = updateRoomRow(state.db, roomId, {
      maxInteractions: body.max_interactions,
      isPaused: body.is_paused,
      isFinished: body.is_finished,
    })
    // Reachable only if the row vanished between the access check and here.
    if (updated === null) throw roomNotFound()
    return c.json(toRoom(updated))
  })

  // ---------------------------------------------------------------------------
  // Pause / resume
  // ---------------------------------------------------------------------------

  /**
   * Pausing stops the room *and* the turn already running in it.
   *
   * The flag alone would only prevent the next round; agents mid-response would
   * keep generating, and their replies would land in a room the user believes
   * is stopped. `interruptRoom` is what makes the button mean what it says.
   */
  routes.post('/rooms/:room_id/pause', async (c) => {
    const roomId = intPathParam(c, 'room_id')
    ensureRoomAccessFor(c, state.db, roomId)

    const room = updateRoomRow(state.db, roomId, { isPaused: true })
    if (room === null) throw roomNotFound()

    await state.orchestrator.interruptRoom(roomId)
    logger.info(`Room ${roomId} paused`)
    return c.json(toRoom(room))
  })

  routes.post('/rooms/:room_id/resume', (c) => {
    const roomId = intPathParam(c, 'room_id')
    ensureRoomAccessFor(c, state.db, roomId)

    const room = updateRoomRow(state.db, roomId, { isPaused: false })
    if (room === null) throw roomNotFound()

    logger.info(`Room ${roomId} resumed`)
    return c.json(toRoom(room))
  })

  // ---------------------------------------------------------------------------
  // Destructive, admin only
  // ---------------------------------------------------------------------------

  // No `ensureRoomAccess` on either: `requireAdmin` has already run, and
  // Python's handlers take no identity at all — they go straight to the
  // service, which is what also tears down the room's warm agent sessions.
  routes.delete('/rooms/:room_id', requireAdmin, async (c) => {
    const roomId = intPathParam(c, 'room_id')
    const deleted = await state.agents.deleteRoomWithCleanup(state.db, roomId)
    if (!deleted) throw roomNotFound()
    return c.json({ message: 'Room deleted successfully' })
  })

  routes.delete('/rooms/:room_id/messages', requireAdmin, async (c) => {
    const roomId = intPathParam(c, 'room_id')
    const cleared = await state.agents.clearRoomMessagesWithCleanup(state.db, roomId)
    // Python conflates the two cases here, and says so in its `detail`: an
    // empty room and a missing one answer identically.
    if (!cleared) throw new HttpError(404, 'Room not found or no messages to delete')
    return c.json({ message: 'All messages cleared successfully' })
  })

  return routes
}
